File: vendor_frontend/frontend/src/models/performance.ts
/**
 * Module 4 – Vendor Performance Management
 * TypeScript interface/model definitions.
 *
 * Excludes backend/database code. Swapping this mock implementation
 * for real HTTP/REST API calls will require zero changes to UI interfaces.
 */

export interface DeliveryPerformance {
  id?: number;
  poNumber: string;
  vendorName: string;
  expectedDeliveryDate: string;
  actualDeliveryDate?: string;
  delayDays?: number;
  deliveryStatus: 'Early' | 'On-Time' | 'Delayed';
  remarks?: string;
}

export interface QualityEvaluation {
  id?: number;
  poNumber: string;
  vendorName: string;
  inspectionDate: string;
  materialQuality: number;          // 1-5
  packagingQuality: number;         // 1-5
  quantityAccuracy: number;         // 1-5
  specificationCompliance: number;  // 1-5
  productDefects: string;
  overallQualityRating: number;      // derived, e.g. average of above
  inspectorRemarks: string;
}

export interface CommunicationLog {
  id?: number;
  poNumber: string;
  vendorName: string;
  messageSentTime: string;
  vendorResponseTime?: string;
  responseDurationHours?: number;   // computed difference
  communicationStatus: 'Awaiting Response' | 'Responded' | 'SLA Breach';
  remarks?: string;
}

export interface ServiceRating {
  id?: number;
  poNumber: string;
  vendorName: string;
  professionalism: number;          // 1-5
  customerSupport: number;          // 1-5
  documentationQuality: number;     // 1-5
  flexibility: number;              // 1-5
  communicationEffectiveness: number; // 1-5
  issueResolution: number;          // 1-5
  overallServiceRating: number;     // derived
  comments: string;
}

export interface VendorPerformanceMetrics {
  vendorName: string;
  onTimeDeliveryRate: number;       // percentage (0-100)
  delayedDeliveryCount: number;
  avgQualityRating: number;         // 1-5 stars
  avgResponseTimeHours: number;
  orderCompletionRate: number;      // percentage (0-100)
  overallPerformanceScore: number;  // 0-100 (weighted aggregate)
}

export interface VendorRanking {
  rankPosition: number;
  vendorName: string;
  category: string;
  overallScore: number;             // 0-100
  deliveryScore: number;            // 0-100
  qualityScore: number;             // 0-100
  communicationScore: number;       // 0-100
  serviceScore: number;             // 0-100
}

File: backend/vendoriq-backend/app/schemas/performance.py
from datetime import datetime, date
from pydantic import BaseModel, Field


class VendorPerformanceCreate(BaseModel):
    vendor_id: str
    purchase_order_id: str | None = None
    record_date: date

    # Delivery
    expected_delivery_date: date | None = None
    actual_delivery_date: date | None = None
    on_time_delivery: bool | None = None
    delivery_delay_days: int = 0
    delivery_remarks: str | None = None

    # Quality
    material_quality: float | None = Field(default=None, ge=0, le=5)
    packaging_quality: float | None = Field(default=None, ge=0, le=5)
    quantity_accuracy: float | None = Field(default=None, ge=0, le=5)
    specification_compliance: float | None = Field(default=None, ge=0, le=5)
    product_defects: str | None = None
    quality_rating: float | None = Field(default=None, ge=0, le=5)
    quality_remarks: str | None = None

    # Communication
    message_sent_time: datetime | None = None
    vendor_response_time: datetime | None = None
    response_time_hours: float | None = None
    communication_status: str | None = None
    communication_remarks: str | None = None

    # Service
    professionalism: float | None = Field(default=None, ge=0, le=5)
    customer_support: float | None = Field(default=None, ge=0, le=5)
    documentation_quality: float | None = Field(default=None, ge=0, le=5)
    flexibility: float | None = Field(default=None, ge=0, le=5)
    communication_effectiveness: float | None = Field(default=None, ge=0, le=5)
    issue_resolution_time_hours: float | None = None
    issue_resolution: float | None = Field(default=None, ge=0, le=5)
    service_rating: float | None = Field(default=None, ge=0, le=5)
    service_comments: str | None = None

    order_completed: bool = False
    remarks: str | None = None


class VendorPerformanceOut(VendorPerformanceCreate):
    id: str
    recorded_by: str | None = None
    created_at: datetime

    class Config:
        from_attributes = True


class VendorPerformanceSummary(BaseModel):
    vendor_id: str
    total_orders: int
    on_time_deliveries: int
    delayed_deliveries: int
    on_time_delivery_rate: float
    avg_quality_rating: float | None
    avg_response_time_hours: float | None
    avg_issue_resolution_time_hours: float | None
    order_completion_rate: float
